### PR TITLE
Split getindex and view apart. Closes #38.

### LINKED ISSUE #38: Should indexing return copies?
When I started development of this package, it was somewhat of a testbed for new indexing behaviors.  I tried to make the package rather opinionated on how I believed indexing should work.  One of those opinions was that indexing should return views by default.

Base Julia has caught up (and surpassed) AxisArrays in terms of APL indexing, but it hasn't caught the view train yet.  Should AxisArrays take a step back and match behaviors?

### PR DESCRIPTION
**Pull request by timholy**
Ref #38

### CODE REVIEW

File: src/indexing.jl [hunk 1]
```diff
@@ -70,44 +28,63 @@ end
...
 # Setindex is so much simpler. Just assign it to the data:
-Base.setindex!(A::AxisArray, v, idxs::Idx...) = (A.data[idxs...] = v)
+@inline Base.setindex!(A::AxisArray, v, idxs::Idx...) = (A.data[idxs...] = v)
 
 ### Fancier indexing capabilities provided only by AxisArrays ###
-Base.getindex(A::AxisArray, idxs...) = A[to_index(A,idxs...)...]
-Base.setindex!(A::AxisArray, v, idxs...) = (A[to_index(A,idxs...)...] = v)
+@inline Base.getindex(A::AxisArray, idxs...) = A[to_index(A,idxs...)...]
+@inline Base.setindex!(A::AxisArray, v, idxs...) = (A[to_index(A,idxs...)...] = v)
+# Deal with lots of ambiguities here
```
Comment (mbauman) [inline src/indexing.jl]: We should probably do the same sort of thing with `view` that we did for `getindex`: only define `view(::T, ::Any...)` in the standard library.  I wish it'd be easier to specify "fallback" methods.

File: src/indexing.jl [hunk 2]
```diff
@@ -70,44 +28,63 @@ end
...
     end
 end
 
+@inline function Base.getindex(A::AxisArray, idxs::Idx...)
+    AxisArray(A.data[idxs...], reaxis(A, idxs...))
+end
+
+# To resolve ambiguities, we need several definitions
+@inline function Base.view{T,N}(A::AxisArray{T,N}, idxs::Vararg{Idx,N})
+    AxisArray(view(A.data, idxs...), reaxis(A, idxs...))
```
Comment (mbauman) [inline src/indexing.jl]: :+1:

File: src/indexing.jl [hunk 3]
```diff
@@ -1,65 +1,23 @@
...
-
-# When we index with non-vector arrays, we *add* dimensions. This isn't
-# supported by SubArray currently, so we instead return a copy.
-# TODO: we probably shouldn't hack Base like this, but it's so convenient...
-if VERSION < v"0.5.0-dev"
-    @inline Base.index_shape_dim(A, dim, i::AbstractArray{Bool}, I...) = (sum(i), Base.index_shape_dim(A, dim+1, I...)...)
-    @inline Base.index_shape_dim(A, dim, i::AbstractArray, I...) = (size(i)..., Base.index_shape_dim(A, dim+1, I...)...)
-end
-@generated function Base.getindex(A::AxisArray, I::Union{Idx, AbstractArray{Int}}...)
+@generated function reaxis(A::AxisArray, I::Idx...)
```
Comment (mbauman) [inline src/indexing.jl]: :100: This makes a lot of sense.

File: src/indexing.jl (hunk 1 above)
Comment (timholy) [inline src/indexing.jl]: Nice cleanup in Base!

**Pull request merged**